New resource: `azurerm_api_management_custom_domain`

sirlatrom opened the pull request:
Closes #3058.

The approach used is to modify the API Management service (corresponding to an `azurerm_api_management` resource) by updating an existing instance. A note is added to the documentation in both that and the new resource telling that you should not configure the custom domains in _both_ resources, as that would likely create conflicts.

Semantically, if you delete an `azurerm_api_management_custom_domain`, the `hostname_configurations` field will be set to be empty (null) when the API Management service instance is updated. There should not be more than one `azurerm_api_management_custom_domain` resource for an `azurerm_api_management`, but I have not found a way to enforce this using the Terraform Provider SDK.

---

Includes:
  - [x] Implementation
  - [x] Acceptance tests
  - [x] Website docs

gd-asharov commented:
Hi,
Do you have a plan when this feature will be completed?

sirlatrom commented:
> Hi,
> Do you have a plan when this feature will be completed?

You can vote for the issue to be prioritized by giving a :+1: reaction to the top-level issue description that this PR should solve: https://github.com/terraform-providers/terraform-provider-azurerm/issues/3058#issue-421658657.

Whenever a maintainer has time to look through the proposed changes and additions, this PR can move forward.

sirlatrom commented:
Rebased just in case.

sirlatrom commented:
@manicminer Is there any chance of this getting reviewed before 2.30.0?

manicminer commented on website/docs/r/api_management_custom_domain.html.markdown:
```diff
@@ -0,0 +1,152 @@
+
+* `host_name` - (Required) The Hostname to use for the legacy Developer Portal.
+
+* `certificate` - (Optional) The Base64 Encoded Certificate. (Mutually exlusive with `key_vault_id`.)
+
+* `certificate_password` - (Optional) The password associated with the certificate provided above.
+
+* `key_vault_id` - (Optional) The ID of the Key Vault Secret containing the SSL Certificate, which must be should be of the type application/x-pkcs12.
+
+* `negotiate_client_certificate` - (Optional) Should Client Certificate Negotiation be enabled for this Hostname? Defaults to false.
+
+---
+
+A `proxy` block supports the following:
+
+* `host_name` - (Required) The Hostname to use for the legacy Developer Portal.
+
+* `certificate` - (Optional) The Base64 Encoded Certificate. (Mutually exlusive with `key_vault_id`.)
+
+* `certificate_password` - (Optional) The password associated with the certificate provided above.
+
+* `default_ssl_binding` - (Optional) Is the certificate associated with this Hostname the Default SSL Certificate? This is used when an SNI header isn't specified by a client. Defaults to false.
+
+* `key_vault_id` - (Optional) The ID of the Key Vault Secret containing the SSL Certificate, which must be should be of the type application/x-pkcs12.
+
+* `negotiate_client_certificate` - (Optional) Should Client Certificate Negotiation be enabled for this Hostname? Defaults to false.
+
+---
+
+A `scm` block supports the following:
+
+* `host_name` - (Required) The Hostname to use for the SCM domain.
+
+* `certificate` - (Optional) The Base64 Encoded Certificate. (Mutually exlusive with `key_vault_id`.)
+
+* `certificate_password` - (Optional) The password associated with the certificate provided above.
+
+* `key_vault_id` - (Optional) The ID of the Key Vault Secret containing the SSL Certificate, which must be should be of the type application/x-pkcs12.
+
+* `negotiate_client_certificate` - (Optional) Should Client Certificate Negotiation be enabled for this Hostname? Defaults to false.
```
Since these are all the same basic set of attributes, I think we should compress these similarly to how they are in the [api_management resource docs](https://www.terraform.io/docs/providers/azurerm/r/api_management.html#host_name). The description for `host_name` could be something like "The hostname to use for the corresponding API endpoint".

manicminer reviewed (changes requested):
Hi @sirlatrom, thanks for this PR! I have a few minor edits and suggestions inline, and I think some more test coverage is needed.

The `testAccAzureRMApiManagementCustomDomain_basic` test looks good, but there is no test that sets any of the custom domains. I'd suggest adding a `testAccAzureRMApiManagementCustomDomain_complete` test having at least two types configured, and then a `testAccAzureRMApiManagementCustomDomain_update` test which first creates a resource with no domains (basic), then adds some domains (complete) and lastly removes then (back to basic), with import steps in between.

If we can expand the tests and fix up the other edits, this looks great to merge!

manicminer commented on azurerm/internal/services/apimanagement/tests/api_management_custom_domain_resource_test.go:
```diff
@@ -0,0 +1,177 @@
+		resp, err := conn.Get(ctx, resourceGroup, serviceName)
+		if err != nil {
+			if utils.ResponseWasNotFound(resp.Response) {
+				// TODO: Is this an error case? If Custom Domains is destroyed, it should not destroy the API Management Service, so what are we testing for here?
+				// If the entire configuration is torn down, including the API Management Service, then it, too, will not be found, and returning nil is correct.
+				return nil
```
Since the CheckDEstroy func is run after a destroy, the APIM service is probably gone too, so it's fine to return nil here.

manicminer replied:
Since the CheckDestroy func is run after a destroy, the APIM service is probably gone too, so it's fine to return nil here.

```suggestion
				return nil
```

sirlatrom commented:
> 
> 
> Hi @sirlatrom, thanks for this PR! I have a few minor edits and suggestions inline, and I think some more test coverage is needed.
> 
> The `testAccAzureRMApiManagementCustomDomain_basic` test looks good, but there is no test that sets any of the custom domains. I'd suggest adding a `testAccAzureRMApiManagementCustomDomain_complete` test having at least two types configured, and then a `testAccAzureRMApiManagementCustomDomain_update` test which first creates a resource with no domains (basic), then adds some domains (complete) and lastly removes then (back to basic), with import steps in between.
> 
> If we can expand the tests and fix up the other edits, this looks great to merge!

@manicminer I've addressed the code comments but have yet to write a `testAccAzureRMApiManagementCustomDomain_update` test. I'm not actually sure if it's actually required to have at least one domain configured - it might be. I've updated the basic test to set two domains of different types. I'm not too familiar with the update type of test.

sirlatrom commented:
Wow, turns out the success I thought I'd had with my first attempt was a false positive, and it really did nothing. I've added several commits and have now verified that it actually does what it is supposed to now.

manicminer commented:
@sirlatrom I haven't tested your recent edits, but it makes sense at a glance.

When testing it seemed like having at least one domain wasn't required, so I figure it makes sense to be able to have a resource with no domains configured - thus enabling removal of existing configured domains?

Given that, it's useful to have a test that configures no domains ("basic"), another that configures some ("complete"), and an "update" test that applies a series of configurations ("basic", then "complete", then "basic") to test that moving from no domains, to some domains, and back to no domains, works as expected.

sirlatrom commented:
> 
> 
> @sirlatrom I haven't tested your recent edits, but it makes sense at a glance.
> 
> When testing it seemed like having at least one domain wasn't required, so I figure it makes sense to be able to have a resource with no domains configured - thus enabling removal of existing configured domains?
> 
> Given that, it's useful to have a test that configures no domains ("basic"), another that configures some ("complete"), and an "update" test that applies a series of configurations ("basic", then "complete", then "basic") to test that moving from no domains, to some domains, and back to no domains, works as expected.

@manicminer Thanks for your feedback and patience with this PR.

The intention (and current implementation in this PR) with regard to being able to remove existing configured domains is that if you've once created an `azurerm_api_management_custom_domain` resource, when you destroy it the custom domains will be removed again. However, I agree it makes sense to validate adding or removing different types of domains in an update test. I'll try and get that done within the next couple of days.

jackofallops commented on azurerm/internal/services/apimanagement/api_management_custom_domain_resource.go:
```diff
@@ -0,0 +1,325 @@
+		return fmt.Errorf("finding API Management (API Management %q / Resource Group %q): %s", name, resourceGroup, err)
+	}
+
+	if d.IsNewResource() {
+		if existing.ServiceProperties.HostnameConfigurations != nil {
+			return tf.ImportAsExistsError(apiManagementCustomDomainResourceName, *existing.ID)
+		}
+	}
```
I think we'll have issue here. If `ServiceProperties` is `nil` we'll crash, and I believe that in a normal response `HostnameConfiguration` isn't nil, so this will always trigger.

jackofallops commented on azurerm/internal/services/apimanagement/api_management_custom_domain_resource.go:
```diff
@@ -0,0 +1,325 @@
+		return fmt.Errorf("cannot read ID for Custom Domain (API Management %q / Resource Group %q)", name, resourceGroup)
+	}
+
+	d.SetId(*read.ID)
```
I suspect we'll need a custom ID for this resource, since this is just the ID of the api_management resource?

sirlatrom replied:
Probably, yes. How about appending `/customDomains/default` to the ID of the API Management service?

sirlatrom replied:
I've gone ahead and appended `/customDomains/default` to the ID, not sure if it needs to be added/assumed any more places.

jackofallops commented on azurerm/internal/services/apimanagement/api_management_custom_domain_resource.go:
```diff
@@ -0,0 +1,325 @@
+		},
+
+		Schema: map[string]*schema.Schema{
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"api_management_name": azure.SchemaApiManagementName(),
```
I think we should use the ID of the api_management resource here, rather than the full schema.  The custom_domain resource is a meta resource that modifies an existing api_management resource, so should be intrinsically linked to it. (we can then get the name and resource group from the ID)

sirlatrom replied:
All the other subresources refer to the APIM by resource_group_name and api_management_name, but if you think direct ID is better (i.e. `api_management_id`) I'll change it to that.

sirlatrom replied:
By looking closer, the SDK client also expects resource group name and name parameters, meaning it would bring additional parsing with it if we switch to `api_management_id`.

katbyte replied:
@sirlatrom - i agree with @jackofallops, we can use the id and parse out the resource group and name with a custom parse function (which this resource should be switched over to anyways) for example see pardse.CdnEndpointID

sirlatrom replied:
I've done it inside the Read function for now. If you think it warrants its own parse function I can make that during the week.

sirlatrom replied:
@katbyte I've made the parse function in e61e67d.

jackofallops commented on azurerm/internal/services/apimanagement/api_management_custom_domain_resource.go:
```diff
@@ -0,0 +1,325 @@
+		// Iterate through old state to find sensitive props not returned by API.
+		// This must be done in order to avoid state diffs.
+		// NOTE: this information won't be available during times like Import, so this is a best-effort.
+		snakeCaseConfigType := azure.ToSnakeCase(string(config.Type))
+		if valsRaw, ok := d.GetOk(snakeCaseConfigType); ok {
+			vals := valsRaw.([]interface{})
+			for _, val := range vals {
+				oldConfig := val.(map[string]interface{})
+
+				if oldConfig["host_name"] == *config.HostName {
+					output["certificate_password"] = oldConfig["certificate_password"]
+					output["certificate"] = oldConfig["certificate"]
+				}
+			}
+		}
```
Since we're going to the trouble of switching on `config.Type` I think we should make this a helper func and move its use to the cases. It will involve a few more LoC, but will be more explicit about what we're attempting to get from config/state, and should help guard against changes to the structure in the API responses.

sirlatrom commented:
> Hi @sirlatrom 
> Thanks for this, I've noted a few changes / issues below.
> Aside from those, was there a reason you chose this as a separate resource over extending the `azrure_api_management` resource with an additional `Optional` schema for this config? (The custom ID note below wouldn't be needed in that approach)

Yes, the sole purpose is to be able to create an `azurerm_key_vault_access_policy` in between the `azurerm_api_management` and this resource. Otherwise, the APIM's System Assigned Identity won't be able to get access to the certificates in the Key Vault.

sirlatrom commented:
I plan to go through your comments this week, hopefully tomorrow CEST.

sirlatrom commented:
Would it be better, at least in the website docs, to split out the access policies into two separate `azurerm_key_vault_access_policy` resources? That is how I would use it in practice.

sirlatrom commented:
Rebased after conflict arose from main branch.

sirlatrom commented:
@katbyte Is there any chance you'd be able to have a look at this during the week?

manicminer commented on azurerm/internal/services/apimanagement/api_management_custom_domain_resource.go:
```diff
@@ -0,0 +1,323 @@
+	}
+
+	if d.IsNewResource() {
+		if existing.ServiceProperties != nil && existing.ServiceProperties.HostnameConfigurations != nil {
```
As per @jackofallops' comment, we should maybe also check for an empty slice at `existing.ServiceProperties.HostnameConfigurations`

```suggestion
		if existing.ServiceProperties != nil && existing.ServiceProperties.HostnameConfigurations != nil && len(*existing.ServiceProperties.HostnameConfigurations) > 0 {
```

manicminer commented on website/docs/r/api_management_custom_domain.html.markdown:
```diff
@@ -0,0 +1,174 @@
+}
+
+resource "azurerm_api_management_custom_domain" "example" {
+  resource_group_name = azurerm_resource_group.example.name
+  api_management_name = azurerm_api_management.example.name
```
```suggestion
  api_management_id = azurerm_api_management.example.id
```

manicminer commented on website/docs/r/api_management_custom_domain.html.markdown:
```diff
@@ -0,0 +1,174 @@
+
+The following arguments are supported:
+
+* `api_management_name` - (Required) The name of the API Management service to configure Custom Domain for. Changing this forces a new API Management Custom Domain to be created.
+
+* `resource_group_name` - (Required) The name of the Resource Group where the API Management resource exists. Changing this forces a new API Management Custom Domain to be created.
```
```suggestion
* `api_management_id` - (Required) The ID of the API Management service for which to configure Custom Domains. Changing this forces a new API Management Custom Domain resource to be created.
```

manicminer commented on azurerm/internal/services/apimanagement/tests/api_management_custom_domain_resource_test.go:
```diff
@@ -0,0 +1,280 @@
+%s
+
+resource "azurerm_api_management_custom_domain" "test" {
+  api_management_name = azurerm_api_management.test.name
+  resource_group_name = azurerm_resource_group.test.name
```
```suggestion
  api_management_id = azurerm_api_management.test.id
```

manicminer commented on azurerm/internal/services/apimanagement/api_management_custom_domain_resource.go:
```diff
@@ -0,0 +1,323 @@
+	log.Printf("[INFO] preparing arguments for API Management Custom domain creation.")
+
+	apiManagementID := d.Get("api_management_id").(string)
+	id, err := azure.ParseAzureResourceID(apiManagementID)
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	serviceName := id.Path["service"]
```
Could we capture this in a custom parse function? e.g. `parse.ApiManagementID`

sirlatrom replied:
Done in 7c1e893d967d0f500fc6196e9a5431000efb9212.

manicminer reviewed (changes requested):
Hi @sirlatrom, sorry for the delay in re-reviewing. Some additional comments inline, and could you add an update test to change/remove the custom domains? Once we have this we should be ok to merge. Thanks!

sirlatrom reviewed (commented):
> Hi @sirlatrom, sorry for the delay in re-reviewing. Some additional comments inline, and could you add an update test to change/remove the custom domains? Once we have this we should be ok to merge. Thanks!

Added update test in fa8a2b2fdf8032f5d087f441ca40b73d50fd321d. Please note that if `hostname_configurations` is specified in the API object, Azure will automatically add an entry for the always-on, built-in proxy domain on `<api_management_service_name>.azure-api.net` - hence the changes to the test configurations.

sirlatrom commented:
@manicminer Is this good to go?

manicminer commented:
@sirlatrom I'm still working on the tests, currently trying to track down a spurious import error. Unfortunately it's slow going since each APIM resource takes about 45 mins to stand up, and yesterday the API was timing out for GET requests.

sirlatrom commented:
> 
> 
> @sirlatrom I'm still working on the tests, currently trying to track down a spurious import error. Unfortunately it's slow going since each APIM resource takes about 45 mins to stand up, and yesterday the API was timing out for GET requests.

APIM is indeed a pain to iterate on... If you need a second pair of eyes on anything, just reach out either here or in Slack.

manicminer commented:
@sirlatrom Could you fix up the vendoring issue with your recent rebase? Thanks!

sirlatrom commented:
> 
> 
> @sirlatrom Could you fix up the vendoring issue with your recent rebase? Thanks!

I believe it's in good shape now: https://travis-ci.org/github/terraform-providers/terraform-provider-azurerm/builds/738292335

manicminer commented:
@sirlatrom Thanks, appreciated. I haven't forgotten about this PR; I've spent several days trying to get a single test to run but being hit by various API error states.

sirlatrom commented:
@manicminer I appreciate how frustrating APIM is to work with, but is there any chance you have made any progress on this PR since last week, or is there perhaps something I can do to move it forward?

manicminer commented:
**Test Results**

`TestAccDataSourceAzureRMApiManagement_`
![Screenshot 2020-11-05 at 14 23 27](https://user-images.githubusercontent.com/251987/98253104-8a2efc80-1f72-11eb-9dea-c1aa028e7f08.png)

`TestAccAzureRMApiManagementCustomDomain_`
![Screenshot 2020-11-05 at 16 05 21](https://user-images.githubusercontent.com/251987/98265224-bfdae200-1f80-11eb-8752-5dd13e7dffe5.png)

jackofallops reviewed (approved):
Thanks @manicminer & @sirlatrom - This LGTM now 👍

manicminer commented:
Thanks @jackofallops :+1:

Latest test results:

![Screenshot 2020-11-11 at 10 13 31](https://user-images.githubusercontent.com/251987/98799104-9699d600-2406-11eb-98d2-564205d90941.png)
